Misc/NEWS.d/next/Library/2019-06-05-14-47-06.bpo-37166.cDr9zg.rst

```diff
@@ -0,0 +1,2 @@
+inspect.findsource() no longer triggers IndexError when co_firstlineno is
+larger than the size of the linecache entry
```

Comment:
Would you mind how that can happen? I understand that the issue can happen if the source file is modified after the code is loaded in Python, right?

Reply:
https://bugs.python.org/issue37166 describes the broad issue.

In our case, we compile the application into an executable package (.par). It may not contain the sources. When the executable is ran, inspect may pick up a source file from the current working directory instead.

This is a common occurrence, as developers tend to run the executable from the root of the repository, and the filenames will always align - and the skew in file lengths between the executable and the working directory is likely.

The following would reproduce the issue closer to the real-life scenario:

_a.py:_
```python
import inspect








def foo(): pass

inspect.getsource(foo)
```

```bash
python3 -m compileall -b a.py
black a.py  # to remove the empty lines
python3 a.pyc  # Index error
```

In the real-life scenario the equivalent of a.pyc could be an executable built in an arbitrary point in time, and distributed into the development environment.

Reply:
I suggest:

:func:`inspect.findsource` no longer triggers :exc:`IndexError` when ``co_firstlineno`` is
larger than the size of the :mod:`linecache` entry. It can happen when the source file
is modified after the code object creation.